msprime/species_tree.py

```diff
@@ -0,0 +1,834 @@
+#
+# Copyright (C) 2020 University of Oxford
+#
+# This file is part of msprime.
+#
+# msprime is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# msprime is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with msprime.  If not, see <http://www.gnu.org/licenses/>.
+#
+"""
+Module responsible for parsing species trees.
+"""
+import re
+
+import msprime
+
+
+class Tree(object):
+    """
+    The tree object, consisting of multiple edges. Trees are checked for
+    whether they are ultrametric; only ultrametric trees are allowed.
+    Species IDs are stored as node IDs of extant edges.
+    """
+    def __init__(self, newick_string):
+        self.newick_string = newick_string
+        self.edges = []
+
+    def get_newick_string(self):
+        return self.newick_string
+
+    def parse_newick_string(self):
+        working_newick_string = self.newick_string
+        number_of_internal_nodes = 0
+        number_of_edges = 0
+
+        # Remove comments from the tree string.
+        pattern = re.compile("\\[.*?\\]")
+        hit = "placeholder"
+        while hit is not None:
+            hit = pattern.search(working_newick_string)
+            if hit is not None:
+                working_newick_string = working_newick_string.replace(hit.group(0), "")
+
+        # Check whether a branch above the root is present, and if so, remove it.
+        if (working_newick_string[0:2] == "((" and working_newick_string[-1] == ")"
+                and working_newick_string[-2] != ")"):
+            level = 0
+            newick_string_tail_start_pos = 0
+            newick_string_tail = ""
+            for pos in range(len(working_newick_string)):
+                if working_newick_string[pos] == "(":
+                    level += 1
+                if working_newick_string[pos] == ")":
+                    level -= 1
+                if level == 1 and pos > 1:
+                    newick_string_tail_start_pos = pos
+                    newick_string_tail = working_newick_string[pos+1:]
+                    break
+            if newick_string_tail.count(",") == 0:
+                nstsp = newick_string_tail_start_pos
+                working_newick_string = working_newick_string[1:nstsp+1]
+
+        # Parse the bifurcating part of the tree.
+        err = 'It appears that the tree string does not include branch lengths!'
+        assert ":" in working_newick_string, err
```

Comment:
We should raise a ValueError here rather than an assert, I think. I think asserts are for internal sanity checks ("this assumption I'm making should definitely be true"), not input validation. So, this would be changed to:
```python
if ":" not in working_newick_string:
    raise ValueError("Species tree strings must include branch lenghts")
```

Reply:
There's a bunch of other places with the same pattern, I won't comment on them.

Reply:
Thanks, I didn't know that. Should now be fixed in my fork.